lambdas/aip.py

```diff
@@ -185,42 +193,81 @@ def _get_aip_s3_inventory(self) -> pd.DataFrame:
                 f"S3 Inventory data not found for S3 key: '{self.s3_key}'"
             )
 
-        return inventory_df
+        # index by file key and return
+        return inventory_df.set_index("key")
 
     @staticmethod
     def _decode_base64_sha256(base64_checksum: str | bytes) -> str:
         binary_checksum = base64.b64decode(base64_checksum)
         return binascii.hexlify(binary_checksum).decode("ascii")
 
-    def _get_aip_file_checksums(self) -> dict[str, str]:
+    def _get_aip_file_checksums(
+        self, num_workers: int = CONFIG.checksum_num_workers
+    ) -> dict[str, str]:
         """Retrieve checksums for all files listed in Bagit manifest.
 
         If a SHA256 checksum does not exist, generate one by copying the object onto
-        itself.
+        itself.  This process is performed in parallel via threads by the worker function
+        'process_file_worker' which updates a local dictionary of file-to-checksum.
         """
         if self.manifest_df is None:
             raise ValueError("Bagit manifest data not found")
         if self.s3_inventory is None:
             raise ValueError("S3 Inventory data not found")
 
         file_checksums = {}
-        s3_inventory_indexed = self.s3_inventory.set_index("key")
+        file_checksums_lock = Lock()
 
-        for _, row in self.manifest_df.iterrows():
+        def process_file_worker(row: pd.Series) -> None:
             filepath = row.filepath
-            logger.debug(f"Getting checksum for AIP file: {filepath}")
-
             s3_uri = f"{self.s3_uri}/{filepath}"
-            inventory_row = s3_inventory_indexed.loc[f"{self.s3_key}/{filepath}"]
+            inventory_row = self.s3_inventory.loc[  # type: ignore[union-attr]
+                f"{self.s3_key}/{filepath}"
+            ]
 
-            if "SHA256" not in inventory_row["checksum_algorithm"]:
+            # retrieve or generate SHA256
+            if inventory_row["checksum_algorithm"] != "SHA256":
                 base64_checksum = self.s3_client.generate_checksum_for_object(s3_uri)
             else:
                 base64_checksum = self.s3_client.get_checksum_for_object(s3_uri)
 
+            # convert base64 form to ascii
             checksum = self._decode_base64_sha256(base64_checksum)
-            logger.debug(f"Filepath '{filepath}', checksum '{checksum}'")
-            file_checksums[filepath] = checksum
+            logger.debug(f"AIP file: '{filepath}', checksum: '{checksum}'")
+
+            # save file:checksum dictionary
+            with file_checksums_lock:
+                file_checksums[filepath] = checksum
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
+            # prepare futures of workers
+            futures = [
+                executor.submit(process_file_worker, row)
+                for _, row in self.manifest_df.iterrows()
+            ]
+
+            total_futures = len(futures)
```

Comment:
Is the number of futures equivalent to the number of files in the AIP / rows in the manifest?

Reply:
That's correct!  Each file in the manifest whether we are generating and/or retrieving the checksum will be it's own "task" (future) to complete.